Mathlib/Analysis/Convex/Radon.lean

```diff
@@ -49,3 +60,124 @@ theorem radon_partition (h : ¬ AffineIndependent 𝕜 f) :
+    · rw [show ⋂ i, F i = ⋂ i ∈ Set.univ, F i by simp]
+      apply h_inter Set.univ
+      rw [Set.ncard_univ]
+      exact hn.le
+      /- Construct a family of vectors indexed by `ι` such that the vector corresponding to `i : ι`
+      is an arbitrary element of the intersection `⋂ i : ι, F i` which *does not lie in `F i`*. -/
+    · let a (i : ι) : E := Set.Nonempty.some (s := ⋂₀ (F '' (Set.univ \ {i}))) <| by
+        let ι' := Set.univ \ {i}
+        let F' : ι' → Set E := fun j ↦ F j
+        rw [show ⋂₀ (F '' (Set.univ \ {i})) = ⋂ i, F' i by simp [ι']]
+        apply hk (F := F')
+        · exact fun i ↦ h_convex ↑i
+        · intro J hJ_card
+          rw [show (⋂ i ∈ J, F' i) = (⋂ i ∈ Subtype.val '' J, F i) by simp]
+          apply h_inter
+          exact le_trans Set.ncard_image_le hJ_card
+        · rw [Nat.card_coe_set_eq ι', Set.ncard_diff_singleton_of_mem]
+          · rw [Set.ncard_univ]
+            omega
+          · trivial
+        · exact Subtype.finite
+      /- This family of vectors is not affine independent because the number of them exceeds the
+      dimension of the space. -/
+      have h2 : ¬AffineIndependent 𝕜 a := by
+        have : Fintype ι := by exact Fintype.ofFinite ι -- for instance inferring
+        rw [← finrank_vectorSpan_le_iff_not_affineIndependent 𝕜 a (n := (k - 1))]
+        · exact (Submodule.finrank_le (vectorSpan 𝕜 (Set.range a))).trans (Nat.le_pred_of_lt h_card)
+        · rw [Fintype.card_eq_nat_card]; omega
+      /- Use `Convex.radon_partition` to conclude there is a subset `I` of `ι` and a point `p : E`
+      which lies in the convex hull of either `a '' I` or `a '' Iᶜ`. We claim that `p ∈ ⋂₀ F`. -/
+      obtain ⟨I, p, h4_I, h4_Ic⟩ := Convex.radon_partition h2
+      use p
+      apply Set.mem_iInter_of_mem
+      intro i
+      /- It suffices to show that for any set `F i` in a subcollection `I` of `ι`, that the convex
+      hull of `a '' Iᶜ` is contained in `F i`. -/
+      suffices ∀ I : Set ι, i ∈ I → (convexHull 𝕜) (a '' Iᶜ) ⊆ F i by
+        by_cases h : i ∈ I
+        · exact this I h h4_Ic
+        · apply this Iᶜ h; rwa [compl_compl]
+      /- Given any subcollection `I` of `ι` containing `i`, because `F i` is convex, we need only
+      show that `a j ∈ F i` for each `j ∈ Iᶜ`. -/
+      intro I hi
+      rw [Convex.convexHull_subset_iff (h_convex i)]
+      rintro - ⟨j, hj, rfl⟩
+      /- Since `j ∈ Iᶜ` and `i ∈ I`, we conclude that `i ≠ j`, and hence by the definition of `a`:
+      `a j ∈ ⋂ F '' (Set.univ \ {j}) ⊆ F i`. -/
+      apply Set.mem_of_subset_of_mem (s₁ := ⋂₀ (F '' (Set.univ \ {j})))
+      · apply sInter_subset_of_mem
+        use i
+        simp only [mem_diff, Set.mem_univ, mem_singleton_iff, true_and, and_true]
+        exact fun h' ↦ hj (h' ▸ hi)
+      · apply Set.Nonempty.some_mem
+
+/-- The set version of `Convex.helly_theorem`. -/
+theorem Convex.helly_theorem_set (F : Set (Set E)) {hF_fin : Set.Finite F}
```

Comment:
```suggestion
theorem Convex.helly_theorem_set {F : Set (Set E)} (hF_fin : Set.Finite F)
```

`F` can be deduced from the type of `hF_fin`, so it should be the implicit one.

(A more idiomatic way of stating this pair of hypotheses is `(F : Finset (Set E))`. In fact, I would argue that many of the occurrences of `Set` in this file could be replaced by `Finset` and the theorems would become easier to use. One benefit of this approach is that you could replace every occurrence of `ncard` with the better-behaved `Finset.card`. However, this would require significant changes, so it's okay to leave it as is for now.)

Reply:
I tried to use `Finset`, but it requires `DecidableEq` in the part of proof where we reason about "all indexes but one".

Reply:
It's on my to-do list to finish my review here, but I was also going to suggest `Finset` among other things. You can get around the need for `DecodableEq` by invoking `classical` in the proof.

Reply:
Can you give me some kind of intuition here? When is `Finset` preferable to `Set.Finite`?

Reply:
If you can package two hypotheses `F : Set β` and `hF_fin : Set.Finite F` into one hypothesis `F : Finset β`, that almost always makes the theorem easier to use.

I don't think that `Set.Finite` and `ncard` should be used at all in this file.